File: STACMR-matlab/staMRFIT.m
function [p, datafit, fits, maxbad, times] = staMRFIT (data, varargin)
% function [p, datafit, fits] = staMRFIT (data, varargin)
% fits the MR model
% nsample = no. of Monte Carlo samples (about 10000 is good)
% data = data structure (cell array or general)
% E = optional partial order model e.g. E={[1 2] [3 4 5]} indicates that
% condition 1 <= condition 2 and condition 3 <= condition 4 <= condition 5
% shrink is parameter to control shrinkage of covariance matrix (if input is not stats form);
% 0 = no shrinkage; 1 = diagonal matrix; -1 = calculate optimum
% returns:
% p = empirical p-value
% datafit = observed fit of monotonic (1D) model
% fits = nsample vector of fits of Monte Carlo samples (it is against this
% distribution that datafit is compared to calculate p)
% *************************************************************************
% Last modified: 15 January 2017
% *************************************************************************
%
if iscell(data)
    if isstruct(data{1})
        y = data; 
    else
        y = data; % cell array format
    end
else
    y = gen2cell(data); % convert from general format
end

nsample = 1;
E = {};
shrink = -1;
reverse = 0;
proc = -1;
mrTol = 0;
ranseed = -1;
showStatus = 0;

for i = 1 : 2 : length(varargin)-1
    name = varargin{i}; 
    value = varargin{i+1};
    switch name
        case {'nsamples', 'nsample', 'ns', 'n'} 
            nsample = value; % number of samples
        case {'partial', 'part', 'p', 'order', 'ord', 'o'}
            E = value; % partial order
        case {'shrink', 's'}
            shrink = value; % shrink value
        case {'reverse', 'rev'}
            reverse = value; % reverse value
        case {'workers', 'work', 'w', 'proc'}
            proc = value; % no. of workers
        case {'mrTol', 'tol', 't'}
            mrTol = value; % approximate solution option
        case {'ranseed','ran','seed'}
            ranseed = value; % random number seed
        case {'showstatus','show','sh'}
            showStatus = value; % showstatus option
    end
end
if isempty(E)
    disp('Warning: Partial order undefined.');
end
if ~iscell(E)
    E = adj2cell(E); % convert from adjacency matrix form
end
[p, datafit, fits, maxbad, times] = jMRfits(nsample, y, E, shrink, reverse, proc, mrTol, ranseed, showStatus);

File: STACMR-R/staMRFIT.R
staMRFIT <- function (data=NULL, partial = list(), nsample=1, shrink=-1) {
# input:
  # nsample = no. of Monte Carlo samples (about 10000 is good)
  # data = data structure (cell array or general)
  # partial = optional partial order model e.g. E={[1 2] [3 4 5]} indicates that
  # condition 1 <= condition 2 and condition 3 <= condition 4 <= condition 5
  # default = none (empty)
  # shrink is parameter to control shrinkage of covariance matrix (if input is not stats form);
  # 0 = no shrinkage; 1 = diagonal matrix; -1 = calculate optimum, default = -1
# output:
  # p = empirical p-value
  # datafit = observed fit of partial order model
  # fits = nsample vector of fits of Monte Carlo samples (it is against this
  # distribution that datafit is compared to calculate p)
  # pars = nvar list of bootstrap means nsample x ncond
  # *************************************************************************
  # converted from matlab 7 February 2018
  # bootstrap parameter estimates added 24 March 2020
  # *************************************************************************

  if (is(data,"data.frame")) {
    y = gen2list (data) # convert from general format to list format
  } else {y = data} 
  
  tol <- 10e-6
  if (missing(partial)) {partial = list()}
  if (missing(shrink)) {shrink = -1}
  
  nvar = length(y[[1]])
  ngroup = length(y); ncond = length(y[[1]][[1]][1,]) * ngroup
  
  if (!is.list(partial)) {partial = adj2list(partial)} # convert from adjacency matrix to list
  
  output = jMRfits(nsample, y, partial, shrink); # call java program

  output$fits[which(output$fits <= tol)] = 0;
  output$datafit[which(output$datafit <= tol)] = 0;
  
  # unpack bootstrap means
  z = array(0,dim=c(nvar,ncond,nsample))
  for (isample in 1:nsample) {
    z[,,isample] = .jevalArray(output$pars[[isample]],simplify=T)
  }
  a = vector("list", nvar)
  z = aperm(z,c(3,2,1))
  for (ivar in 1:nvar){
    a[[ivar]] = z[,,ivar]
  }
  output$pars = a;
  
  return (output)
}